File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/objects/attributes/Attribute.java
package org.saar.lwjgl.opengl.objects.attributes;

import org.lwjgl.opengl.GL20;

public class Attribute implements IAttribute {

    private final AttributeLinker linker;
    private final int attributeIndex;

    public Attribute(AttributeLinker linker, int attributeIndex) {
        this.linker = linker;
        this.attributeIndex = attributeIndex;
    }

    @Override
    public void enable() {
        GL20.glEnableVertexAttribArray(this.attributeIndex);
    }

    @Override
    public void disable() {
        GL20.glDisableVertexAttribArray(this.attributeIndex);
    }

    @Override
    public void link(int stride, int offset) {
        this.linker.link(this.attributeIndex, stride, offset);
    }

    @Override
    public int getBytesPerVertex() {
        return this.linker.getBytes();
    }
}

File: pe-lwjgl-binding/src/main/java/org/saar/lwjgl/opengl/attribute/Attribute.kt
package org.saar.lwjgl.opengl.attribute

import org.lwjgl.opengl.GL20
import org.saar.lwjgl.opengl.attribute.divisor.AttributeDivisor
import org.saar.lwjgl.opengl.attribute.pointer.AttributePointer

class Attribute(
    private val index: Int,
    private val linker: AttributePointer,
    private val divisor: AttributeDivisor,
) : IAttribute {

    override fun enable() = GL20.glEnableVertexAttribArray(this.index)

    override fun disable() = GL20.glDisableVertexAttribArray(this.index)

    override fun link(stride: Int, offset: Int) {
        this.linker.point(this.index, stride, offset)
        this.divisor.divide(this.index)
    }

    override val bytesPerVertex get() = this.linker.bytesPerVertex
}